# If-Statements with User-Defined Functions

In [1]:
import pandas as pd
import numpy as np
import os

In [2]:
# define function

def add_num(x,y):
    sum = x + y
    return sum

In [3]:
add_num(9, 11)

20

In [2]:
# create a path

path = r'C:\Users\apiny\Documents\CareerFoundry\Data Immersion\4_Intro to Programming\Instacart Basket Analysis'

In [3]:
# import orders_products_merged dataframe

df_ords_prods_merged = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged.pkl'))

In [4]:
# check the output

df_ords_prods_merged.shape

(32404859, 14)

In [6]:
# check the output

df_ords_prods_merged.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  aisle_id  department_id  prices _merge  
0         Soda        77              7     9.0   both  
1         Soda        77              7     9.0   both  
2         Soda        77              7     9.0   both  
3         Soda        77              7     9.0   both  
4         Soda        77              7     9.0   both

In [7]:
# create a subset 1 million rows

df = df_ords_prods_merged[:1000000]

In [8]:
df.shape

(1000000, 14)

In [9]:
# define a function

def price_label(row):
    if row['prices'] <= 5:
        return 'low-range product'
    elif (row['prices'] > 5) and (row['prices'] <= 15):
        return 'mid-range product'
    elif row['prices'] > 15:
        return 'high-range product'
    else: return 'Not enough data'

In [10]:
# apply the function

df['price_range'] = df.apply(price_label, axis = 1)

C:\Users\apiny\AppData\Local\Temp\ipykernel_22800\1179148871.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['price_range'] = df.apply(price_label, axis = 1)


In [11]:
# check the value in a new column

df['price_range'].value_counts(dropna = False)

price_range
mid-range product    756450
low-range product    243550
Name: count, dtype: int64

In [12]:
df['prices'].max()

14.8

# If-Statements with the loc() Function

In [13]:
df_ords_prods_merged.loc[df_ords_prods_merged['prices'] > 15, 'price_range_loc'] = 'high-range product'

In [14]:
df_ords_prods_merged.loc[(df_ords_prods_merged['prices'] <= 15) & (df_ords_prods_merged['prices'] > 5), 'price_range_loc'] = 'mid-range product'

In [15]:
df_ords_prods_merged.loc[df_ords_prods_merged['prices'] <= 5, 'price_range_loc'] = 'low-range product'

In [17]:
# check the output

df_ords_prods_merged['price_range_loc'].value_counts(dropna = False)

price_range_loc
mid-range product     21860860
low-range product     10126321
high-range product      417678
Name: count, dtype: int64

# If-Statements with For-Loops

In [18]:
# print the frequency of the “orders_day_of_week” column (which day most orders take place)

df_ords_prods_merged['orders_day_of_week'].value_counts(dropna = False)

orders_day_of_week
0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: count, dtype: int64

0 means Saturday. This value has the highest frequency, meaning Saturday is the busiest day.
The 4 value has the lowest frequency. A value of 4, refers to Wednesday, meaning Wednesday is the slowest day for Instacart app orders.

In [19]:
# create a result list

result = []

for value in df_ords_prods_merged['orders_day_of_week']:
    if value == 0:
        result.append('Busiest day')
    elif value == 4:
        result.append('Least busy')
    else:
        result.append('Regularly busy')

In [19]:
result

['Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Busiest day',
 'Regularly busy',
 'Reg

In [20]:
# create a new column 'busiest day'

df_ords_prods_merged['busiest_day'] = result

In [21]:
# check the output

df_ords_prods_merged['busiest_day'].value_counts(dropna = False)

busiest_day
Regularly busy    22416875
Busiest day        6204182
Least busy         3783802
Name: count, dtype: int64

In [22]:
df_ords_prods_merged.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  aisle_id  department_id  prices _merge    price_range_loc  \
0         Soda        77              7     9.0   both  mid-range product   
1         Soda        77              7     9.0   both  mid-range product   
2         Soda        77              7     9.0   both  mid-range product   
3         Soda        77              7     9.0   both  mid-range product   
4         Soda        77              7     9.0   both  mid-range product   

      busiest_day  
0  Regularly busy  
1  Regularly busy  
2  Regularly busy  
3      Least busy  
4      Least busy

# Exercise 4.7

## Step 1 create the “price_label” and “busiest_day” columns

In [23]:
# rename 'price_range_loc' to 'price_label'

df_ords_prods_merged.rename(columns = {'price_range_loc' : 'price_label'}, inplace = True)

In [24]:
# drop '_merge' column

df_ords_prods_merged = df_ords_prods_merged.drop(columns = ['_merge'])

In [25]:
# check the output

df_ords_prods_merged.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  aisle_id  department_id  prices        price_label  \
0         Soda        77              7     9.0  mid-range product   
1         Soda        77              7     9.0  mid-range product   
2         Soda        77              7     9.0  mid-range product   
3         Soda        77              7     9.0  mid-range product   
4         Soda        77              7     9.0  mid-range product   

      busiest_day  
0  Regularly busy  
1  Regularly busy  
2  Regularly busy  
3      Least busy  
4      Least busy

## Step 2 create a new column shown two busiest days and two slowest days

In [26]:
# create a result list for busiest days
# two busiest days are the value 0 and 1 in orders_day_of_week column, which mean Saturday and Sunday
# two slowest days are the value 3 and 4 in orders_day_of_week column, which mean Tuesday and Wednesday

result = []

for value in df_ords_prods_merged['orders_day_of_week']:
    if value in (0, 1):
        result.append('busiest days')
    elif value in (3, 4):
        result.append('slowest days')
    else: 
        result.append('regularly busy')

In [27]:
result

['regularly busy',
 'slowest days',
 'slowest days',
 'slowest days',
 'slowest days',
 'regularly busy',
 'busiest days',
 'busiest days',
 'busiest days',
 'slowest days',
 'busiest days',
 'regularly busy',
 'regularly busy',
 'busiest days',
 'busiest days',
 'regularly busy',
 'regularly busy',
 'slowest days',
 'slowest days',
 'slowest days',
 'slowest days',
 'slowest days',
 'slowest days',
 'busiest days',
 'busiest days',
 'busiest days',
 'regularly busy',
 'regularly busy',
 'busiest days',
 'regularly busy',
 'regularly busy',
 'busiest days',
 'regularly busy',
 'regularly busy',
 'regularly busy',
 'slowest days',
 'regularly busy',
 'slowest days',
 'busiest days',
 'busiest days',
 'regularly busy',
 'slowest days',
 'slowest days',
 'regularly busy',
 'regularly busy',
 'busiest days',
 'busiest days',
 'regularly busy',
 'busiest days',
 'busiest days',
 'slowest days',
 'regularly busy',
 'busiest days',
 'busiest days',
 'busiest days',
 'busiest days',
 'busiest 

In [28]:
# create a new column 'busiest days'

df_ords_prods_merged['busiest_days'] = result

In [29]:
df_ords_prods_merged.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  aisle_id  department_id  prices        price_label  \
0         Soda        77              7     9.0  mid-range product   
1         Soda        77              7     9.0  mid-range product   
2         Soda        77              7     9.0  mid-range product   
3         Soda        77              7     9.0  mid-range product   
4         Soda        77              7     9.0  mid-range product   

      busiest_day    busiest_days  
0  Regularly busy  regularly busy  
1  Regularly busy    slowest days  
2  Regularly busy    slowest days  
3      Least busy    slowest days  
4      Least busy    slowest days

## Step 3 check the value from step 2 for accuracy

In [30]:
# check the output of busiest days

df_busiest_days = df_ords_prods_merged.loc[df_ords_prods_merged['orders_day_of_week'].isin([0,1])]

In [31]:
df_busiest_days.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
6     550135        1             7                   1                  9   
7    3108588        1             8                   1                 14   
8    2295261        1             9                   1                 16   
10   2968173       15            15                   1                  9   
13   2715276       15            21                   1                  9   

    days_since_prior_order  product_id  add_to_cart_order  reordered  \
6                     20.0         196                  1          1   
7                     14.0         196                  2          1   
8                      0.0         196                  4          1   
10                     7.0         196                  2          0   
13                     7.0         196                  1          1   

   product_name  aisle_id  department_id  prices        price_label  \
6          Soda        77              7     9.0  mid-range product   
7          Soda        77              7     9.0  mid-range product   
8          Soda        77              7     9.0  mid-range product   
10         Soda        77              7     9.0  mid-range product   
13         Soda        77              7     9.0  mid-range product   

       busiest_day  busiest_days  
6   Regularly busy  busiest days  
7   Regularly busy  busiest days  
8   Regularly busy  busiest days  
10  Regularly busy  busiest days  
13  Regularly busy  busiest days

In [32]:
df_busiest_days['busiest_days'].value_counts(dropna = False)

busiest_days
busiest days    11864412
Name: count, dtype: int64

In [33]:
df_busiest_days.shape

(11864412, 16)

11864412 rows have the value of 0 and 1 in 'orders_day_of_week' column and all of them have shown 'busiest days' in busiest_days' column.

In [34]:
# check the output of slowest days

df_slowest_days = df_ords_prods_merged.loc[df_ords_prods_merged['orders_day_of_week'].isin([3,4])]

In [35]:
df_slowest_days.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   
9   2550362        1            10                   4                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   
9                    30.0         196                  1          1   

  product_name  aisle_id  department_id  prices        price_label  \
1         Soda        77              7     9.0  mid-range product   
2         Soda        77              7     9.0  mid-range product   
3         Soda        77              7     9.0  mid-range product   
4         Soda        77              7     9.0  mid-range product   
9         Soda        77              7     9.0  mid-range product   

      busiest_day  busiest_days  
1  Regularly busy  slowest days  
2  Regularly busy  slowest days  
3      Least busy  slowest days  
4      Least busy  slowest days  
9      Least busy  slowest days

In [36]:
df_slowest_days['busiest_days'].value_counts(dropna = False)

busiest_days
slowest days    7624336
Name: count, dtype: int64

In [37]:
df_slowest_days.shape

(7624336, 16)

7624336 rows have the value of 3 and 4 in 'orders_day_of_week' column and all of them have shown 'slowest days' in busiest_days' column.

## Step 4 Create a new column “busiest_period_of_day”

In [38]:
# check the frequency of the “order_hour_of_day” column

df_ords_prods_merged['order_hour_of_day'].value_counts(dropna = False)

order_hour_of_day
10    2761760
11    2736140
14    2689136
15    2662144
13    2660954
12    2618532
16    2535202
9     2454203
17    2087654
8     1718118
18    1636502
19    1258305
20     976156
7      891054
21     795637
22     634225
23     402316
6      290493
0      218769
1      115700
5       87961
2       69375
4       53242
3       51281
Name: count, dtype: int64

In [39]:
df_ords_prods_merged['order_hour_of_day'].describe()

count    3.240486e+07
mean     1.342515e+01
std      4.246380e+00
min      0.000000e+00
25%      1.000000e+01
50%      1.300000e+01
75%      1.600000e+01
max      2.300000e+01
Name: order_hour_of_day, dtype: float64

Considering the frequecy of 'order_hour_of_day', we can identify the busiest hours of the day into three groups; 
“Most orders” between 10 - 16, 
“Average orders” between 8 - 9 and 17 - 19, 
and “Fewest orders” between 0 - 7 and 20 - 23

In [40]:
# create a result list for busiest period of day

result = []

for value in df_ords_prods_merged['order_hour_of_day']:
    if value in (10,11,12,13,14,15,16):
        result.append('Most orders')
    elif value in (8,9,17,18,19):
        result.append('Average orders')
    else:
        result.append('Fewest orders')

In [41]:
result

['Average orders',
 'Fewest orders',
 'Most orders',
 'Fewest orders',
 'Most orders',
 'Fewest orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Fewest orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Fewest orders',
 'Most orders',
 'Mo

In [42]:
# create a new column

df_ords_prods_merged['busiest_period_of_day'] = result

In [43]:
# check the output

df_ords_prods_merged.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  aisle_id  department_id  prices        price_label  \
0         Soda        77              7     9.0  mid-range product   
1         Soda        77              7     9.0  mid-range product   
2         Soda        77              7     9.0  mid-range product   
3         Soda        77              7     9.0  mid-range product   
4         Soda        77              7     9.0  mid-range product   

      busiest_day    busiest_days busiest_period_of_day  
0  Regularly busy  regularly busy        Average orders  
1  Regularly busy    slowest days         Fewest orders  
2  Regularly busy    slowest days           Most orders  
3      Least busy    slowest days         Fewest orders  
4      Least busy    slowest days           Most orders

In [44]:
# check the accuracy

df_busiest_period = df_ords_prods_merged.loc[df_ords_prods_merged['order_hour_of_day'].isin([10,11,12,13,14,15,16])]

In [45]:
df_busiest_period['busiest_period_of_day'].value_counts(dropna = False)

busiest_period_of_day
Most orders    18663868
Name: count, dtype: int64

In [46]:
df_ords_prods_merged['order_hour_of_day'].isin([10,11,12,13,14,15,16]).value_counts(dropna = False)

order_hour_of_day
True     18663868
False    13740991
Name: count, dtype: int64

In [47]:
df_average_period = df_ords_prods_merged.loc[df_ords_prods_merged['order_hour_of_day'].isin([8,9,17,18,19])]

In [48]:
df_average_period['busiest_period_of_day'].value_counts(dropna = False)

busiest_period_of_day
Average orders    9154782
Name: count, dtype: int64

In [49]:
df_ords_prods_merged['order_hour_of_day'].isin([8,9,17,18,19]).value_counts(dropna = False)

order_hour_of_day
False    23250077
True      9154782
Name: count, dtype: int64

## Step 5 print the frequency for 'busiest_period_of_day' column

In [50]:
df_ords_prods_merged['busiest_period_of_day'].value_counts(dropna = False)

busiest_period_of_day
Most orders       18663868
Average orders     9154782
Fewest orders      4586209
Name: count, dtype: int64

## Step 7 export a dataframe as a pickle file

In [51]:
# export a dataframe

df_ords_prods_merged.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_label.pkl'))